### DISCUSSION TITLE
My collection of sub-queries is growing and I'm discovering more and more reusability needs

### DISCUSSION THREAD

**Discussion by Madd0g** [cozodb/cozo#153] [Q&A]
So in a bunch of parts of my code, I'm already doing quite a bit of rule-glueing, like adding sub-rules to a query, to further filter the results. Currently it's on the string-level.

I'm wondering if anyone has conceived or conceptualized a query builder or query merger-er? I've seen previous discussions about views and storing inline rules, I think it's a bit related. In other engines, some of the things I'm doing could be achieved with functions/views.

Thanks for sharing!

**Comment by zh217** [comment]
Take a look at https://github.com/cozodb/pycozo/blob/main/pycozo/builder.py and https://github.com/pegesund/cozodb-lisp
Views inside the database is also an option, but it comes with quite a lot of complexity so we want to think clearly beforehand.

**Reply by Madd0g** [reply]
thank you, I found the test for the [pycozo builder](https://github.com/cozodb/pycozo/blob/main/pycozo/test_builder.py), it's good, the kind of example I was hoping to see.

Seeing it laid out like this, I am able to read it pretty fluently, I'm not sure I want to switch to something like this, it's too low level for me and obscures the original language, while I'm still learning to be fluent at it.

I think the builder pattern is the way to go, I'm thinking how it would look like on a slightly higher level.

Thanks
